mlir/lib/Quantum/Transforms/AdjointPatterns.cpp

```diff
@@ -0,0 +1,224 @@
+            return bvm.lookupOrDefault(yield->getOperand(0));
+        }
+        else {
+            rewriter.insert(i->clone(bvm));
+        }
+    }
+    assert(false && "quantum.yield must present in the adjoint region");
+}
+
+struct AdjointSingleOpRewritePattern : public mlir::OpRewritePattern<AdjointOp> {
+    using mlir::OpRewritePattern<AdjointOp>::OpRewritePattern;
+
+    /// In essence, we build a map from values mentiond in the source data flow to the values of the
+    /// program where quantum control flow is reversed. Most of the time, there is a 1-to-1
+    /// correspondence with a notable exception caused by `insert`/`extract` API asymetry.
+    mlir::LogicalResult matchAndRewrite(AdjointOp adjoint,
+                                        mlir::PatternRewriter &rewriter) const override
+    {
+        LLVM_DEBUG(dbgs() << "Adjointing the following:\n" << adjoint << "\n");
+        Location loc = adjoint.getLoc();
+        MLIRContext *ctx = adjoint.getContext();
+        assert(adjoint.getRegion().hasOneBlock());
+
+        // First, copy the classical computations directly to the target POI and build the classical
+        // value mapping dictionary.
+        auto classicalMapping = ({
+            IRMapping out;
+            Block &b = adjoint.getRegion().front();
+            for (auto i = b.begin(); i != b.end(); i++) {
+                if (i->getName().getStringRef().find("quantum") != std::string::npos) {
+                    /* Skip quantum operations */
+                }
+                else {
+                    LLVM_DEBUG(dbgs() << "classical operation: " << *i << "\n");
+                    rewriter.insert(i->clone(out));
+                }
+            }
+            out;
+        });
+
+        // Next, compute and copy the reversed quantum computation flow. The classical dependencies
+        // such as gate parameters or qubit indices are already available in `classicalMapping`.
+        auto quantumMapping = ({
+            llvm::DenseMap<Value, Value> out;
+            auto query = [&out](Value key) -> Value {
+                LLVM_DEBUG(dbgs() << "  querying: " << key << "\n");
+                auto val = out[key];
+                LLVM_DEBUG(dbgs() << "    result: " << val << "\n");
+                return val;
+            };
+            auto update = [&out](Value key, Value val) -> void {
+                LLVM_DEBUG(dbgs() << "  updating: " << key << "\n");
+                LLVM_DEBUG(dbgs() << "    to: " << val << "\n");
+                out[key] = val;
+            };
+            Block &b = adjoint.getRegion().front();
+            auto rb = std::make_reverse_iterator(b.end());
+            auto re = std::make_reverse_iterator(b.begin());
+            for (auto i = rb; i != re; i++) {
+                LLVM_DEBUG(dbgs() << "operation: " << *i << "\n");
+                if (YieldOp yield = isInstanceOf<YieldOp>(*i)) {
+                    assert(yield.getOperands().size() == 1);
```

Comment:
Will there ever be a case where this isn't true?

Reply:
I would say that ideally we want the individual qubits to appear here in the future whenever possible.

Reply:
Yes, indeed. If we support single qubits, then we might have more than one value here.